libr/util/json.c

```diff
@@ -0,0 +1,391 @@
+#else
+#define R_JSON_REPORT_ERROR(msg, p) do { (void)(msg); (void)(p); } while (0)
+#endif
+
+static RJson *json_new(void) {
+	return R_NEW0 (RJson);
+}
+
+static RJson *create_json(RJsonType type, const char *key, RJson *parent) {
+	RJson *js = json_new ();
+	if (!js) {
+		return NULL;
+	}
+	js->type = type;
+	js->key = key;
+	if (!parent->children.last) {
+		parent->children.first = parent->children.last = js;
+	} else {
+		parent->children.last->next = js;
+		parent->children.last = js;
+	}
+	parent->children.length++;
+	return js;
+}
+
+R_API void r_json_free(RJson *js) {
+	if (!js) {
+		return;
+	}
+	if (js->type == R_JSON_OBJECT || js->type == R_JSON_ARRAY) {
+		RJson *p = js->children.first;
+		RJson *p1;
+		while (p) {
+			p1 = p->next;
+			r_json_free (p);
+			p = p1;
+		}
+	}
+	free (js);
+}
+
+static char *unescape_string(char *s, char **end) {
+	char *p = s;
+	char *d = s;
+	char c;
+	while ((c = *p++)) {
+		if (c == '"') {
+			*d = '\0';
+			*end = p;
+			return s;
+		} else if (c == '\\') {
+			switch (*p) {
+			case '\\':
+			case '/':
+			case '"':
+				*d++ = *p++;
+				break;
+			case 'b':
+				*d++ = '\b';
+				p++;
+				break;
+			case 'f':
+				*d++ = '\f';
+				p++;
+				break;
+			case 'n':
+				*d++ = '\n';
+				p++;
+				break;
+			case 'r':
+				*d++ = '\r';
+				p++;
+				break;
+			case 't':
+				*d++ = '\t';
+				p++;
+				break;
+			case 'u': { // unicode
+				char *ps = p - 1;
+				ut8 high = 0, low = 0;
+				if (r_hex_to_byte (&high, p[1]) || r_hex_to_byte (&high, p[2])
+						|| r_hex_to_byte (&low, p[3]) || r_hex_to_byte (&low, p[4])) {
+					R_JSON_REPORT_ERROR ("invalid unicode escape", p - 1);
+					return NULL;
+				}
+				RRune codepoint = (RRune)high << 8 | (RRune)low;
+				if ((codepoint & 0xfc00) == 0xd800) { // high surrogate; need one more unicode to succeed
+					p += 6;
+					high = low = 0;
+					if (p[-1] != '\\' || *p != 'u'
+							|| r_hex_to_byte (&high, p[1]) || r_hex_to_byte (&high, p[2])
+							|| r_hex_to_byte (&low, p[3]) || r_hex_to_byte (&low, p[4])) {
+						R_JSON_REPORT_ERROR ("invalid unicode surrogate", ps);
+						return NULL;
+					}
+					RRune codepoint2 = (RRune)high << 8 | (RRune)low;
+					if ((codepoint2 & 0xfc00) != 0xdc00) {
+						R_JSON_REPORT_ERROR ("invalid unicode surrogate", ps);
+						return NULL;
+					}
+					codepoint = 0x10000 + ((codepoint - 0xd800) << 10) + (codepoint2 - 0xdc00);
+				}
+				int sz = r_utf8_encode ((ut8 *)d, codepoint);
+				if (!s) {
+					R_JSON_REPORT_ERROR ("invalid codepoint", ps);
+					return NULL;
+				}
+				d += sz;
+				p += 5;
+				break;
+			}
+			default:
+				// leave untouched
+				*d++ = c;
+				break;
+			}
+		} else {
+			*d++ = c;
+		}
+	}
+	R_JSON_REPORT_ERROR ("no closing quote for string", s);
+	return NULL;
+}
+
+static char *skip_block_comment(char *ps) {
+	// ps is at "/* ..."
+	// caller must ensure that ps[0], ps[1] and ps[2] are valid.
+	char *p = ps + 2;
```

Comment:
it will be good to run this code thru radamsa, i think there are other bugs like this in other places

Reply:
Yes, it is checked. The only callsite is in `skip_whitespace()` where the call to this function is enclosed in `if (*p == '/' && p[1] == '*')` so `p[2]` is automatically valid. If you really want to have checks there, I can add them but it would be redundant.

Actually I also thought about fuzzing this api before. It might even be a nice target to run klee on since it's small enough to get some good results.